### PR TITLE
Add grant progress encoder

### LINKED ISSUE #3058: [FEATURE]: History log encoder for grants
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

For creating a history log from a grant, we need an encoder that captures the `failures`.

### Proposed Solution

Create a history log encoder for grants to encode the `failures`.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by JCZuurmond**
## Changes

Add grant progress encoder

### Linked issues

Resolves #3058

### Functionality

- [x] modified existing workflow: `migration-progress`

### Tests

- [x] added unit tests

### CODE REVIEW

File: tests/unit/progress/test_grants.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+@pytest.mark.parametrize(
+    "grant, failure",
+    [
+        (
+            Grant("principal", "DENY", "hive_metastore", "schema"),
+            "Action 'DENY' on DATABASE 'hive_metastore.schema' unmappable to Unity Catalog",
+        ),
+        (
+            Grant("principal", "READ_METADATA", "hive_metastore", "schema", "table"),
```
Comment (JCZuurmond) [inline tests/unit/progress/test_grants.py:48]: What do we think of failing for this grant? As read metadata is not that important?

File: src/databricks/labs/ucx/progress/grants.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+    A failure for a grants implies it cannot be mapped to Unity Catalog.
+    """
+
+    def _encode_record_as_historical(self, record: Grant) -> Historical:
+        historical = super()._encode_record_as_historical(record)
+        failures = []
+        if not record.uc_grant_sql():
+            type_, key = record.this_type_and_key()
+            failures = [f"Action '{record.action_type}' on {type_} '{key}' unmappable to Unity Catalog"]
```
Comment (nfx) [inline src/databricks/labs/ucx/progress/grants.py:19]: this would never be fixed 🤷

Comment (JCZuurmond) [reply src/databricks/labs/ucx/progress/grants.py:19]: So, does that mean we should not visualize it in the migration progress dashboard? That introduces the risk that people will not be away about it.

Comment (JCZuurmond) [reply src/databricks/labs/ucx/progress/grants.py:19]: In theory, people can fix it by rewriting their DENY to GRANT, thus fixing it

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**